### PR TITLE
asymptote 2.61 (new formula)

### PR DESCRIPTION
**Pull request by ellio167**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
new formula for graphics language asymptote

### CODE REVIEW

**Comment by ellio167** [comment]
It was tested and build failed with ‘uses_from_macos’

**Comment by fxcoudert** [comment]
Our CI machines don't have TeX installed:

```
cd doc && /Applications/Xcode.app/Contents/Developer/usr/bin/make asy-latex.pdf
pdflatex asy-latex.dtx
make[1]: pdflatex: No such file or directory
make[1]: *** [asy-latex.pdf] Error 1
make: *** [sty] Error 2
```

Can Asymptote be installed in a way that does not make assumptions about the user's TeX install?

**Comment by ellio167** [comment]
@fxcoudert OK, I think I've got a way to make this all work.  Let me know if there are any tricks I should use to shorten/cleanup the Formula.  Thanks.

**Comment by fxcoudert** [comment]
Taking files from a pre-built archive is not acceptable in Homebrew core. We would rather build without documentation, if that's the choice.

What I'm not sure is: given that asymptote needs a tex installation, and tex installations typically carry asymptote, what is the value of having it in Homebrew?

**Comment by ellio167** [comment]
> Taking files from a pre-built archive is not acceptable in Homebrew core. We would rather build without documentation, if that's the choice.
> 
Fair enough.

> What I'm not sure is: given that asymptote needs a tex installation, and tex installations typically carry asymptote, what is the value of having it in Homebrew?

A good point/question.  My main answer is updates.  TeXLive is only released once a year.  The TeXLive 2019 release ships asymptote-2.49.  The current version of asymptote is 2.61.  There were 14 releases during 2019.  If this is not sufficient to justify inclusion in Homebrew core, I understand.  But, if it is, I'll be happy to workout a build-without-docs Formula.

**Comment by fxcoudert** [comment]
> I'll be happy to workout a build-without-docs formula

Yes, if it's useful, let's go for that!

> TeXLive is only released once a year

I don't know the policy well, but I run `tlmgr update --all` on my laptop every week, and there are updates ever week ;)

**Comment by ellio167** [comment]
> > I'll be happy to workout a build-without-docs formula
> 
> Yes, if it's useful, let's go for that!

Great, I'll work on it.

> 
> > TeXLive is only released once a year
> 
> I don't know the TeXLive policy well, but I run `tlmgr update --all` on my laptop every week, and there are updates ever week ;)

Yes, I do to.  However, these updates are only from CTAN type packages.  

For reference: 
They update arch-dependent binaries only once a year: https://tug.org/mactex/updateschedule.html

> The command line programs must be compiled for individual platforms, and because that is a large undertaking, it only happens once a year.
>
>The style files, class files, fonts, and documentation do not need to be compiled, so they can be distributed as soon as they become available. TeX Live administrators update that part of the distribution daily. To obtain these updates, use TeX Live Utility, installed by MacTeX in /Applications/TeX. Some TUG members update their personal copies just once a year, because the distribution is carefully tested for yearly updates. Other members update every day.

**Comment by ellio167** [comment]
OK, I think I've removed the doc build as cleanly as possible.

**Comment by fxcoudert** [comment]
Hum… sadly that's way to much patching for us, it makes the build fragile and the formula hard to maintain in the long term. I was thinking of `make asy` or `make install-asy`, but even that requires to build things with TeX.

It looks like it's simply not modular enough… sorry

**Comment by ellio167** [comment]
OK, understood.  I'll see if I can work with upstream...

**Comment by ellio167** [comment]
I found a simpler way.  Will this work?

**Comment by ellio167** [comment]
Latest includes adding the manual pdf file so that the runtime "help" command works; also some style fixes from 'brew audit'

File: Formula/asymptote.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  desc "Powerful descriptive vector graphics language"
+  homepage "https://asymptote.sourceforge.io"
+  url "https://downloads.sourceforge.net/project/asymptote/2.61/asymptote-2.61.src.tgz"
+  sha256 "f3ccaa6fde02a2d37fca771563597828530f94821e1c3fe3a5c7d81e7d398d3d"
+
+  depends_on "fftw"
+  depends_on "ghostscript"
+  depends_on "glm"
+  depends_on "gsl"
+  depends_on "readline"
```
Comment (fxcoudert) [inline Formula/asymptote.rb]: Can it run with macOS's libeedit?

Comment (ellio167) [reply Formula/asymptote.rb]: Apparently not... (I see no mention of histedit.h"

Comment (ellio167) [reply Formula/asymptote.rb]: I suggested editline support to upstream.  They have put in a fallback to editline if readline is not available (https://github.com/vectorgraphics/asymptote/commit/a0c44e0b438217ad04eb7833a8acc3a7bd571c0f).  This is in the git repo, and should become available with the next release.

How do you propose we proceed in the mean time?  I could add a comment about removing `depends_on "readline"` in the next release to make sure we remember to make the change...

Comment (fxcoudert) [reply Formula/asymptote.rb]: I say add a comment next to `readline` saying `remove in next release to use macOS editline`.

Comment (ellio167) [reply Formula/asymptote.rb]: done.

**Comment by ellio167** [comment]
@BrewTestBot test this please

**Comment by ellio167** [comment]
```
failed: brew audit asymptote --online --new-formula

Stacktrace

        Error: 2 problems in 1 formula detected
asymptote:
  * C: 11: col 24: Unnecessary spacing detected.
  * 'asymptote' is blacklisted from homebrew/core.
```
Latest push fixes the spacing error (which didn't show up on my local system)

Why is 'asymptote' blacklisted?  Was it because it needed as TeX system for build? We've worked around that here...  I hope the blacklisting can be removed.

**Comment by SMillerDev** [comment]
https://github.com/Homebrew/brew/blob/de72a2310930c3f0903292770b433f197a8d8a6c/Library/Homebrew/missing_formula.rb#L31-L34 says that it's because already shipped as cask.

**Comment by ellio167** [comment]
@SMillerDev Thanks for pointing me to the source.  It is true that it comes as part of MacTeX. But as I say above, MacTeX binaries are updated only once a year.  Asymptote is trending toward more than 10 releases per year.  So, I think it is useful to have an upgrade option through homebrew-core.

**Comment by SMillerDev** [comment]
I don't disagree with you, just let you know the reason. You could make a pull request to remove the blacklisting quoting this PR.

File: Formula/asymptote.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  depends_on "readline" # Remove in next release to use macOS editline
+
+  resource "manual" do
+    url "https://downloads.sourceforge.net/project/asymptote/2.61/asymptote.pdf"
+    sha256 "b5f5848515b6fd6e6d70cebc1b33c45e9d0f00ab7fcf50f71f95e7748c98f5da"
+  end
+
+  def install
+    touch buildpath/"doc/asy-latex.pdf"
+    ENV.deparallelize
```
Comment (MikeMcQuaid) [inline Formula/asymptote.rb]: Is this something that can be reported upstream as a bug? How long does this build take on your machine?

File: Formula/asymptote.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  depends_on "gsl"
+  depends_on "readline" # Remove in next release to use macOS editline
+
+  resource "manual" do
+    url "https://downloads.sourceforge.net/project/asymptote/2.61/asymptote.pdf"
+    sha256 "b5f5848515b6fd6e6d70cebc1b33c45e9d0f00ab7fcf50f71f95e7748c98f5da"
+  end
+
+  def install
+    touch buildpath/"doc/asy-latex.pdf"
```
Comment (MikeMcQuaid) [inline Formula/asymptote.rb]: Can you add a comment explaining why this is needed?

File: Formula/asymptote.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  desc "Powerful descriptive vector graphics language"
+  homepage "https://asymptote.sourceforge.io"
+  url "https://downloads.sourceforge.net/project/asymptote/2.61/asymptote-2.61.src.tgz"
+  sha256 "f3ccaa6fde02a2d37fca771563597828530f94821e1c3fe3a5c7d81e7d398d3d"
+
+  depends_on "fftw"
+  depends_on "ghostscript"
+  depends_on "glm"
+  depends_on "gsl"
+  depends_on "readline" # Remove in next release to use macOS editline
```
Comment (MikeMcQuaid) [inline Formula/asymptote.rb]:
```suggestion
  depends_on "readline" # Remove in >2.61 to use macOS editline
```

File: Formula/asymptote.rb (hunk 2 above)
Comment (ellio167) [reply Formula/asymptote.rb]: With `ENV.deparallelize` it takes about 4:30 minutes.  However, I now see that building without MacTeX fixes the parallelize issue.  So, without `ENV.deparallelize` it takes 1:30 minutes.  I'll remove the `ENV.deparallelize` line.

File: Formula/asymptote.rb (hunk 3 above)
Comment (ellio167) [reply Formula/asymptote.rb]: Will do

File: Formula/asymptote.rb (hunk 4 above)
Comment (ellio167) [reply Formula/asymptote.rb]: will do

**Comment by MikeMcQuaid** [comment]
Thanks so much for your contribution! Without people like you submitting PRs we couldn't run this project. You rock, @ellio167!

**Pull request closed without merging**